src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.State.cs

```diff
@@ -0,0 +1,204 @@
+                var projectRoot = PathUtilities.GetDirectoryName(document.Project.FilePath);
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                return !PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath);
+            }
+
+            public static async Task<State> CreateAsync(TService service, Document document, TextSpan textSpan, CancellationToken cancellationToken)
+            {
+                if (document.Project.FilePath == null)
+                {
+                    return null;
+                }
+
+                if (!textSpan.IsEmpty)
+                {
+                    return null;
+                }
+
+                var workspace = document.Project.Solution.Workspace;
+                if (workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+                {
+                    return null;
+                }
+
+                if (document.IsGeneratedCode(cancellationToken))
+                {
+                    return null;
+                }
+
+                // Bail if we can't get the root namespace for the project, the refactoring depends on it.
+                var rootNamespace = document.Project.DefaultNamespace;
+                if (rootNamespace == null)
+                {
+                    return null;
+                }
+
+                // Ignore linked documents
+                // TODO: this only detect documents linked multiple times, need to figure out how to detect a document is linked, e.g. using <Link>, or shared project, etc.
+                if (IsLinkedDocument(document))
+                {
+                    return null;
+                }
+
+                // Bail if cursor is not at the proper location.
+                (var shouldTriggerRefactoring, var namespaceDeclaration) = 
+                    await service.ShouldPositionTriggerRefactoringAsync(document, textSpan.Start, cancellationToken).ConfigureAwait(false);
+
+                if (!shouldTriggerRefactoring)
+                {
+                    return null;
+                }
+
+                // Namespace can't be renamed if we can't construct a valid qualified identifier from folder names.
+                // In this case, we might still be able to provide refactoring to move file to new location.
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+                var namespaceFromFolders = TryBuildNamespaceFromFolders(service, document.Folders, syntaxFacts);
+                var targetNamespace = namespaceFromFolders == null ? null : ConcatNamespace(rootNamespace, namespaceFromFolders);                
+
+                // namespaceDeclaration == null means the target namespace is global namespace.
```

Comment:
it esems like it woudl be cleaner to just have a boolean for this.

Reply:
Changed code, no longer applies.